Revit_Core_Engine/Query/PanelSurfaces.cs

```diff
@@ -39,6 +40,59 @@ public static partial class Query
         /****              Public methods               ****/
         /***************************************************/
 
+        [Description("Extracts the BHoM-representative location surfaces from a given collection Revit host objects." +
+                     "\nOptimised to regenerate the document only twice for all input elements, not twice per each.")]
+        [Input("hostObjects", "Revit host objects to extract the location surfaces from.")]
+        [Input("discipline", "Engineering discipline based on the BHoM discipline classification.")]
+        [Input("settings", "Revit adapter settings to be used while performing the query.")]
+        [Output("surfaces", "BHoM-representative location surfaces extracted from the input Revit host objects.")]
+        public static Dictionary<ElementId, Dictionary<PlanarSurface, List<PlanarSurface>>> PanelSurfaces(this List<HostObject> hostObjects, Discipline discipline, RevitSettings settings = null)
+        {
+            settings = settings.DefaultIfNull();
+
+            Dictionary<ElementId, Dictionary<PlanarSurface, List<PlanarSurface>>> result = new Dictionary<ElementId, Dictionary<PlanarSurface, List<PlanarSurface>>>();
+
+            List<HostObject> linked = hostObjects.Where(x => x.Document.IsLinked).ToList();
+            List<HostObject> curtains = hostObjects.Where(x => !x.Document.IsLinked && x.ICurtainGrids().Count != 0).ToList();
+            List<HostObject> solids = hostObjects.Except(linked.Union(curtains)).ToList();
+
+            // First process linked elements
+            foreach (HostObject hostObject in linked)
+            {
+                if (hostObject.ICurtainGrids().Count != 0)
+                    result.Add(hostObject.Id, null);
+                else
+                {
+                    IList<ElementId> insertsToIgnore = hostObject.InsertsToIgnore(discipline);
+                    result.Add(hostObject.Id, hostObject.PanelSurfaces_LinkDocument(insertsToIgnore, settings));
+                }
+            }
+
+            if (result.Count(x => x.Value != null) != 0)
+                BH.Engine.Base.Compute.RecordWarning("Pulling panels and openings from Revit link documents is simplified compared to pulling directly from the host document, therefore it may result in degraded output.\n" +
+                                                     "In case of requirement for best possible outcome, it is recommended to open the link document in Revit and pull the elements directly from there.");
+
+            // Return null for curtain panels because their location is extracted differently
+            foreach (HostObject hostObject in curtains)
+            {
+                result.Add(hostObject.Id, null);
+            }
+
+            // Extract location surfaces for all panels at once
+            foreach (var group in solids.GroupBy(x => x.Document)) 
```

Comment:
This is an extra safety measure, just in case it was possible to convert elements from multiple documents (imagine having a few documents open and running some sort of batch query).